### ISSUE TITLE
consumer_openai accepts response_format and silently ignores it

### ISSUE THREAD

**Issue by jeromeetienne** [webai-at-home/webai-at-home#191]
Found by the OpenAI API Conformance Test built in #182, milestone seven. Full results: [`docs/openai_api_conformance.md`](https://github.com/webai-at-home/webai-at-home/blob/main/docs/openai_api_conformance.md).

## What happens

`consumer_openai` accepts `response_format` — both `{"type":"json_object"}` and `{"type":"json_schema", ...}` — returns HTTP 200, and answers with ordinary prose. The field is never read:

```bash
grep -rn "response_format" packages/consumer_openai/src/
```

returns nothing.

## Why it happens

`packages/consumer_openai/src/api/openai_types.ts` sorts request fields into those it reads and those it does not, and says of the second group:

> Every other field an OpenAI client may send — `n`, `logprobs`, `presence_penalty`, `frequency_penalty`, and the rest — is still accepted and then ignored, so the schema deliberately does not refuse unknown fields.

`response_format` has landed in that group. But `n` and `logprobs` *tune* an answer, while `response_format` changes its **shape** — which is what `tools` does, and `tools` is refused rather than ignored. It belongs with `tools` and with the generation controls, not with `n`.

## Why this is a defect rather than a limitation

The rule already written in this codebase, in `packages/consumer_openai/src/api/generation_settings_builder.ts`:

> Being told is better than receiving an answer generated some other way and being told nothing.

A client that asks for `json_object` and receives prose has no way to know its request was dropped. It will call `JSON.parse` on an English sentence.

The comparison run makes the cost concrete. LM Studio serving `llama-3.2-1b-instruct` honours `json_schema` and passes the test. `consumer_openai` in front of that same LM Studio serving that same weights fails it, answering ``  `Hello`  ``.

## An ignored field fails at random, which is why this is easy to miss

Two runs of `structured_output.json_object` against the same endpoint and the same model, minutes apart, with nothing changed between them:

- Run one: PASS — the model happened to answer with bare JSON.
- Run two: WARN — the model happened to wrap the same JSON in a markdown code fence.

Neither outcome had anything to do with the endpoint honouring the request, because it never did. Testing this by hand once and seeing JSON come back would give entirely the wrong impression.

## Suggested fix

Follow the shape already used for `tools` and for the generation controls: decide per task type whether `response_format` can be honoured, honour it where it can be, and refuse the request where it cannot, rather than accepting it and answering some other way.

## Which conformance tests catch it

```bash
npx tsx packages/openai_conformance_test/src/cli.ts --model llm_llama3_2_1b_full --profile structured_output
```

`structured_output.json_schema` fails. `structured_output.json_object` passes or warns depending on luck, as above.

**Comment by jeromeetienne** [comment]
Still valid, checked on `main` at [`6fb2aeb`](https://github.com/webai-at-home/webai-at-home/commit/6fb2aeb). `grep -rn "response_format" packages/consumer_openai/src/` still returns nothing, `ChatCompletionRequestSchema` in `packages/consumer_openai/src/api/openai_types.ts` still declares no `response_format` field, and the comment quoted in the description is still there unchanged.

Here is the plan to fix it.

## The shape the fix follows

`response_format` is treated exactly the way the five generation controls are treated, because the rule that `packages/consumer_openai/src/api/generation_settings_builder.ts` already writes down is the rule this defect breaks. A `response_format` has three possible fates, decided per task type:

- The task type honours it: it is carried to the cluster as the client asked for it.
- The task type does not honour it and the value is this interface's own default, which is `{ "type": "text" }` and an absent or `null` field: it is dropped without a word, because such a client is asking for nothing unusual.
- The task type does not honour it and the value is anything else, so `json_object` and `json_schema`: the request is refused with HTTP 400. Being told is better than receiving an answer generated some other way and being told nothing.

## Milestone 0 — de-risk gate: is there any task type that can honour it today?

The one assumption that would make the rest of the plan wrong is that some task type can already honour `response_format`, in which case the fix would have to carry the field rather than refuse it. Prove or kill that assumption before writing any implementation code.

`packages/protocol/src/task/generation_control_support.ts` already states the rule that decides this: a task type's contract is the intersection of what all of its workers can honour, from step 1 of #196. `task_type_llm_llama3_2_1b_full` can be run by a native worker forwarding to LM Studio, which honours `json_schema`, and by a worker browser tab running `@huggingface/transformers`, which offers no constrained decoding at all. The intersection is empty, so the contract is empty.

The gate is one live run for each of the two workers that can serve `task_type_llm_llama3_2_1b_full`, against the real transport and not a stand-in, asking each for a `json_schema` answer and recording the raw output. Expected outcome: the native worker forwarding to LM Studio honours it, the worker browser tab does not, so no task type honours it and the honest answer for every task type today is a refusal. If the gate says otherwise, the milestones below still hold and the table in milestone 2 simply gets a non-empty entry.

## Milestone 1 — read the field

In `packages/consumer_openai/src/api/openai_types.ts`:

- Add a `ResponseFormatSchema` covering the three values the OpenAI Chat Completions interface allows: `{ "type": "text" }`, `{ "type": "json_object" }`, and `{ "type": "json_schema", "json_schema": { "name": ..., "schema": ..., "strict": ..., "description": ... } }`.
- Add `response_format: ResponseFormatSchema.nullish()` to `ChatCompletionRequestSchema`, beside `tools` and `tool_choice`, because it belongs with the fields that change an answer's shape.
- Rewrite the comment block the issue quotes, so that `response_format` is described alongside `tools` as a field that is read, and so that the list of accepted-and-ignored fields no longer silently contains it.

## Milestone 2 — say which task type honours which response format

Add `packages/protocol/src/task/structured_output_support.ts`, a `StructuredOutputSupport` class mirroring `GenerationControlSupport` file for file:

- A `ResponseFormatName` type, `'json_object' | 'json_schema'`. `text` is not one of them, for the same reason `isStreaming` is not a `GenerationControlName`: it is the default and there is nothing for a consumer to be told about it.
- A `formatsByTaskType: Record<TaskType, readonly ResponseFormatName[]>` table, filled from what milestone 0 observed live rather than from what any engine's documentation claims, with the same per-task-type explanation the generation control table carries.
- `StructuredOutputSupport.honours(taskType, responseFormatName)` and `StructuredOutputSupport.honouredFormats(taskType)`.

Export both the class and the type from `packages/protocol/src/index.ts`, and name the new file in `packages/protocol/src/task/CONTEXT.md`.

This table is the extension point. A task type starts honouring `json_schema` by gaining one entry here, not by any consumer changing.

## Milestone 3 — refuse where it cannot be honoured

Add `packages/consumer_openai/src/api/response_format_reader.ts`, a `ResponseFormatReader` class that applies the three-way rule of the section above, and call it from `_handleChatCompletion` in `packages/consumer_openai/src/http/openai_routes.ts`, next to where the tool declarations and the tool choice are already refused.

Add `OpenaiError.unhonourableResponseFormat(requestedType, modelId, honouredTypes)` to `packages/consumer_openai/src/api/openai_error.ts`, HTTP 400, `param` set to `response_format`, worded the way `unhonourableGenerationControl` and `unsupportedToolDeclarations` are worded: name what was asked for, name the model, and name what that model does honour.

Carrying an honoured `response_format` to the cluster needs a protocol field that does not exist yet, and with an empty table in milestone 2 that carrying path would be code nothing reaches. So it is deliberately left out of this issue: milestone 3 refuses, and the carrying path is written by whichever issue first makes a task type honour a response format, which is when it can be tested against something real. This is a stated cost of the plan, not an oversight.

## Milestone 4 — tests

- Unit tests in `packages/consumer_openai/tests/index.test.ts`: `{ "type": "json_object" }` and `{ "type": "json_schema", ... }` are refused with HTTP 400 and an error body naming `response_format`; `{ "type": "text" }`, `null`, and an absent field are all accepted and answered exactly as they were before; a `response_format` whose shape is wrong is refused by the schema.
- Run the conformance profile again:

  npx tsx packages/openai_conformance_test/src/cli.ts --model llm_llama3_2_1b_full --profile structured_output

  `structured_output.json_schema` goes from `FAIL` to `SKIP`, and `structured_output.json_object` goes from a verdict that changes run to run to a steady `SKIP`, because both tests already read a 4xx as an endpoint declining an optional feature. A steady `SKIP` is the honest report; the `PASS` seen sometimes today was luck.
- Update `docs/openai_api_conformance.md`: the two rows of the results table, and the section titled "`response_format` is accepted and ignored", which becomes a record of a fixed defect rather than a live one.

## Milestone 5 — land it

One commit per milestone, each naming this issue, and the last one carrying `fixes #191`. A comment is posted on this issue as each milestone completes.

**Comment by jeromeetienne** [comment]
## Milestone 0 — de-risk gate: no task type can honour `response_format` today

The assumption under test was that some task type can already honour `response_format`, which would make the fix a carrying path rather than a refusal. The assumption is killed. Here is what was checked, and how.

### Part A — LM Studio honours `json_schema`, live

`lms server start`, then one request straight to LM Studio, no cluster in it:

```bash
curl -s http://localhost:1234/v1/chat/completions -H 'Content-Type: application/json' -d '{
 "model":"llama-3.2-1b-instruct",
 "messages":[{"role":"user","content":"Reply with a greeting object whose greeting is \"hello\"."}],
 "response_format":{"type":"json_schema","json_schema":{"name":"greeting_object","strict":true,"schema":{"type":"object","properties":{"greeting":{"type":"string"}},"required":["greeting"],"additionalProperties":false}}}
}'
```

Raw answer:

```json
{
  "choices": [
    {
      "index": 0,
      "message": { "role": "assistant", "content": "{\n  \"greeting\": \"hello\"\n}", "tool_calls": [] },
      "finish_reason": "stop"
    }
  ],
  "usage": { "prompt_tokens": 20, "completion_tokens": 18, "total_tokens": 38 },
  "system_fingerprint": "llama-3.2-1b-instruct"
}
```

So one of the two workers that can serve `stage_llm_llama3_2_1b_full` — the `@webai/worker-openai` process forwarding to LM Studio — sits in front of an engine that honours `json_schema`.

### Part B — the worker browser tab cannot honour it

The other worker that can serve `stage_llm_llama3_2_1b_full` is the worker browser tab, whose `packages/worker_webpage/web/src/stages/stage_helper_llm_llama3_2_1b_full.ts` generates through `pipeline('text-generation', ...)` from `@huggingface/transformers`, pinned at `4.2.0`.

That library offers no way to ask for a JSON schema or a grammar. This part is a source-level check of the pinned version rather than a live run, and is reported as such, because the finding is that there is no option to pass at all:

```bash
grep -rln "json_schema\|jsonSchema\|GrammarConstrained\|grammar" node_modules/@huggingface/transformers/src/
```

returns nothing, and the one field whose name suggests it, `constraints`, is declared and never read:

```
src/generation/configuration_utils.js:221:    constraints = null;
src/generation/logits_sampler.js:87:  // - *constrained beam-search decoding*: `constraints!=None` or `force_words_ids!=None`
```

The second of those two lines is a comment listing the decoding strategies the upstream Python library has, inside a `getSampler` that goes on to build only a multinomial, a beam search, or a greedy sampler. Nothing reads `constraints`.

### Part C — nothing carries a response format anyway

Neither half of the carrying path exists:

- `@webai/protocol` has no response format field. `GenerationSettingsSchema` carries the five generation controls and `isStreaming`, and nothing else.
- `packages/worker_openai/src/libs/openai_api_client.ts` builds its outgoing request body from `temperature`, `top_p`, `max_tokens`, `stop`, `seed`, and, since [`6fb2aeb`](https://github.com/webai-at-home/webai-at-home/commit/6fb2aeb), `tools`. There is no `response_format` in it, so even the worker sitting in front of LM Studio does not forward one.

### Verdict

`task_type_llm_llama3_2_1b_full` is served by two workers. One of them can honour `json_schema`; the other cannot honour anything of the kind. A task type's contract is the intersection of what all of its workers can honour, from step 1 of #196, so the contract is empty. Every other language-model task type runs only in a worker browser tab on the same library, so every one of them is empty too.

**No task type honours any response format today, so the honest answer for every one of them is a refusal.** The plan holds as posted: the table in milestone 2 is added empty, milestone 3 refuses, and the carrying path stays out of this issue until a task type gains an entry in that table.

Milestone 1 next.

**Comment by jeromeetienne** [comment]
## Milestones 1, 2, and 3

### Milestone 1 — the field is read

[`ae4ec80`](https://github.com/webai-at-home/webai-at-home/commit/ae4ec80). `packages/consumer_openai/src/api/openai_types.ts` gains a `ChatCompletionResponseFormatSchema` covering the three values the OpenAI Chat Completions interface defines — `text`, `json_object`, and `json_schema` with its `name`, `description`, `schema`, and `strict` — and `ChatCompletionRequestSchema` gains `response_format`, declared beside `tools` because both change the shape of an answer rather than tuning it. The comment quoted in this issue's description no longer describes `response_format` as accepted and ignored.

### Milestone 2 — the support table

[`70722eb`](https://github.com/webai-at-home/webai-at-home/commit/70722eb). `packages/protocol/src/task/structured_output_support.ts` adds `StructuredOutputSupport`, mirroring `GenerationControlSupport` file for file: a `ResponseFormatName` of `'json_object' | 'json_schema'`, a `formatsByTaskType` table, and `honours` and `honouredFormats`. `text` is not a `ResponseFormatName`, for the same reason `isStreaming` is not a `GenerationControlName`.

Every entry is empty, and the table's own comment records what the milestone 0 gate measured for each task type rather than what any engine's documentation claims. `packages/protocol/src/index.ts` exports both, and `packages/protocol/src/task/CONTEXT.md` names the new file, its rule, and this issue.

### Milestone 3 — the refusal

[`6a3c795`](https://github.com/webai-at-home/webai-at-home/commit/6a3c795). `packages/consumer_openai/src/api/response_format_reader.ts` adds `ResponseFormatReader`, applying the three-way rule, and `_handleChatCompletion` in `packages/consumer_openai/src/http/openai_routes.ts` calls it next to where the tool declarations and the tool choice are already refused. `OpenaiError.unhonourableResponseFormat` answers HTTP 400 with `param` set to `response_format` and code `unhonourable_response_format`, naming what was asked for, the model, and what that model does produce.

So today:

- `response_format` absent, `null`, or `{ "type": "text" }` — accepted and answered exactly as before.
- `{ "type": "json_object" }` or `{ "type": "json_schema", ... }` — refused with HTTP 400 by every model this server offers.

As the plan said, carrying an honoured `response_format` to the cluster is deliberately not written here: with every entry of `StructuredOutputSupport` empty it would be code nothing reaches, and it belongs to whichever issue first makes a task type produce a shape, which is when it can be tested against something real.

The 58 tests of `npm run test --workspace @webai/consumer-openai` still pass, unchanged. Milestone 4 adds the tests for the new behaviour and re-runs the conformance profile.

**Comment by jeromeetienne** [comment]
## Milestone 4 — tests, and the conformance rerun live

[`3cb4647`](https://github.com/webai-at-home/webai-at-home/commit/3cb4647).

### Unit tests

Three tests in `packages/consumer_openai/tests/index.test.ts`, under a new "The Response Format Of A Request" section:

- The schema reads all three values the OpenAI Chat Completions interface defines, accepts `null`, and refuses a type it does not define as well as a `json_schema` carrying no schema block.
- An absent field, a `null` field, and `{ "type": "text" }` all ask for nothing, so `ResponseFormatReader.read` answers `undefined` for each.
- `{ "type": "json_object" }` and `{ "type": "json_schema", ... }` are both refused over a real HTTP connection with HTTP 400, code `unhonourable_response_format`, `param` `response_format`, and a message naming both the shape asked for and what the model does produce.

`npm run test --workspace @webai/consumer-openai` — 61 passed, 0 failed.

### The conformance rerun, against a real cluster

Not a fixture: the central gateway, `consumer_openai`, and a `@webai/worker-openai` process offering `stage_llm_llama3_2_1b_full` forwarding to a live LM Studio serving `llama-3.2-1b-instruct`, all started for this run.

```bash
npx tsx packages/openai_conformance_test/src/cli.ts --model llm_llama3_2_1b_full --profile structured_output
```

```
Structured Output
  Skipped response_format=json_object — json_object is not supported: HTTP 400, The model
  llm_llama3_2_1b_full cannot produce a response_format of json_object, and this server refuses a
  request it would have to ignore rather than answering it in prose as though no shape had been
  asked for. The response_format types llm_llama3_2_1b_full produces are text only. …
  Skipped response_format=json_schema — json_schema is not supported: HTTP 400, …

Passed: 0
Failed: 0
Skipped: 2
```

`structured_output.json_schema` went from `FAIL` to `SKIP`, and `structured_output.json_object` from a verdict that changed run to run to a steady `SKIP`. The same profile against `llm_qwen3_5_0_8b_full` gives the same two skips.

The `core` profile was run on the same live cluster to check nothing else moved: **Passed 9, Failed 0, Skipped 0, Compatibility 100.0%**. An ordinary request is answered exactly as it was.

### Documentation

`docs/openai_api_conformance.md`: both `structured_output` rows in both `consumer_openai` columns are now ⊘, the per-column totals and percentages are recomputed, the note on what was measured records this second rerun, and the section that read "`response_format` is accepted and ignored" now reads "was accepted and ignored — fixed on 2026-08-16" and says what replaced it.

The paragraph on what the numbers do not say now carries the awkward part plainly: `consumer_openai` running llama rises from 90.0% to 100.0% because a defect was fixed by refusing something rather than by doing it, and fourteen of its thirty-two tests are skipped.

### One thing left untouched

`npm run test --workspaces` fails in `packages/webai_at_home_cli` with `Cannot find package '@webai/openai-conformance-test'`, a missing workspace link in `node_modules`. It fails the same way with every change of this issue stashed, so it predates this work and is left alone. Every other workspace passes.

This closes the issue.

**Comment by jeromeetienne** [comment]
## The one thing left untouched, followed up

The milestone 4 comment above recorded that `npm run test --workspaces` failed in `packages/webai_at_home_cli` with `Cannot find package '@webai/openai-conformance-test'`, and left it alone as something that predates this work. It is now cleared, and it was never a defect in the repository.

The cause was a stale `node_modules` on the machine that ran the tests: `node_modules/@webai/` held a symbolic link for every workspace except `openai-conformance-test`, because `packages/openai_conformance_test` was added after the last `npm install` on that machine. `packages/webai_at_home_cli/src/cli.ts` imports `@webai/openai-conformance-test/dist/cli.js`, and there was nothing for the import to resolve to.

The fix is one command, and it changes no file:

```bash
npm install
```

That recreates the missing link:

```
node_modules/@webai/openai-conformance-test -> ../../packages/openai_conformance_test
```

`git status` is clean afterwards, `package-lock.json` included, so there is nothing to commit.

Afterwards the whole suite passes: `npm run test` runs the documentation link test and every workspace, **423 tests, 0 failures**, `packages/webai_at_home_cli` among them at 12 passed.
